### PR TITLE
Enable pendulum network on zenlink

### PR DESCRIPTION
**Pull request by gianfra-t**
Part of [Tasks/194](https://github.com/pendulum-chain/tasks/issues/194) and [Tasks/211](https://github.com/pendulum-chain/tasks/issues/211).

This changes are required for Pendulum to interact with the Zenlink's UI interface. All of the required hooks are reused form Amplitude.

Also we [add](https://github.com/pendulum-chain/zenlink-interface/pull/2/files#diff-20bf3530498201b648f0f049684a4ddff46081ff0923dde0d7623420c5196b94R25) some assets definitions for Amplitude.

### CODE REVIEW

**Comment by gianfra-t** [comment]
This was tested locally with zombienet, running a Foucoco chain with `CHAIN_ID=2094`. A local version of the squids indexer was also required to test. Only one pair and farm pool was added (PEN<-> DOT).

Swapping and interacting with the pool was working, with the exception of one error in the pools UI probably related to the low block number of the mock chain.

image
image

**Comment by gianfra-t** [comment]
@pendulum-chain/devs this is just an internal review PR, if you think this is fine I will point this to the upstream repo and open the actual pr for zenlink.

File: packages/chain/parachains.json [hunk 1]
```diff
@@ -75,5 +75,30 @@
...
+    "faucets": [],
+    "nativeCurrency": {
+      "name": "Pendulum",
+      "symbol": "PEN",
+      "decimals": 12
+    },
+    "infoURL": "https://pendulumchain.org/",
+    "shortName": "pendulum",
+    "chainId": 2094,
+    "networkId": 200,
```
Comment (ebma) [inline packages/chain/parachains.json]: I think `300` is the `networkId` of Polkadot.

File: packages/config/graph/src/index.ts [hunk 2]
```diff
@@ -20,10 +21,12 @@ export const SQUID_HOST: Record<number | string, string> = {
...
   [ParachainId.BASE]: 'https://api.studio.thegraph.com/query/48211/uniswap-v3-base/version/latest',
   [ParachainId.AMPLITUDE]: `${SQUID_HOST_ENDPOINT}/amplitude-squid/graphql`,
+  [ParachainId.PENDULUM]: `${SQUID_HOST_ENDPOINT}/pendulum-squid/graphql`,
 }
 
 export const ARCHIVE_HOST: Record<number | string, string> = {
   [ParachainId.BIFROST_KUSAMA]: 'https://bifrost.explorer.subsquid.io/graphql',
   [ParachainId.BIFROST_POLKADOT]: 'https://bifrost-polkadot.explorer.subsquid.io/graphql',
   [ParachainId.AMPLITUDE]: 'https://amplitude.explorer.subsquid.io/graphql',
+  [ParachainId.PENDULUM]: 'https://amplitude.explorer.subsquid.io/graphql',
```
Comment (ebma) [inline packages/config/graph/src/index.ts]:
```suggestion
  [ParachainId.PENDULUM]: 'https://pendulum.explorer.subsquid.io/graphql',
```

File: packages/config/router/src/index.ts [hunk 3]
```diff
@@ -127,6 +127,24 @@ export const BASES_TO_CHECK_TRADES_AGAINST: { readonly [chainId: number]: Token[
     WNATIVE[ParachainId.AMPLITUDE],
     KSM[ParachainId.AMPLITUDE],
   ],
+  [ParachainId.PENDULUM]: [
+    WNATIVE[ParachainId.PENDULUM],
+    DOT[ParachainId.PENDULUM],
+    new Token({
+      chainId: ParachainId.PENDULUM,
+      address: '2094-2-262',
+      decimals: 12,
```
Comment (ebma) [inline packages/config/router/src/index.ts]: I think this token has `18` decimals. Which makes it a little more complex for us in other places...

File: packages/currency/src/constants/tokenAddresses.ts
Comment (ebma) [inline packages/currency/src/constants/tokenAddresses.ts:1]: Should we add our `USDT` token to the respective array too? I honestly don't know what this would change in the UI but it made sense.

File: packages/graph-client/src/appolo/client.ts [hunk 4]
```diff
@@ -57,6 +57,11 @@ export const CLIENTS: Record<number | string, ApolloClient<NormalizedCacheObject
...
     cache: new InMemoryCache(),
   }),
+  [ParachainId.PENDULUM]: new ApolloClient({
+    ...DEFAULT_CLIENT_OPTIONS,
+    link: createLink(ParachainId.PENDULUM),
+    cache: new InMemoryCache(),
+  }),
 }
 
 export const ARCHIVE_CLIENTS: Record<number | string, ApolloClient<NormalizedCacheObject>> = {
```
Comment (ebma) [inline packages/graph-client/src/appolo/client.ts:67]: Should we add pendulum here too?

File: packages/token-lists/lists/pendulum.json [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        "name": "Polkadot"
+      },
+      {
+        "networkId": 200,
+        "address": "2094-2-262",
+        "parachainId": 2094,
+        "assetType": 2,
+        "assetIndex": 262,
+        "symbol": "GLMR",
+        "decimals": 10,
```
Comment (ebma) [inline packages/token-lists/lists/pendulum.json]:
```suggestion
        "decimals": 18,
```

File: packages/token-lists/lists/pendulum.json [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        "name": "Moonbeam"
+      },
+      {
+        "networkId": 200,
+        "address": "2094-2-512",
+        "parachainId": 2094,
+        "assetType": 2,
+        "assetIndex": 512,
+        "symbol": "XLM",
+        "decimals": 10,
```
Comment (ebma) [inline packages/token-lists/lists/pendulum.json]:
```suggestion
        "decimals": 12,
```

File: packages/token-lists/lists/pendulum.json [hunk 7]
```diff
@@ -0,0 +1,48 @@
+{
+    "name": "pendulum",
+    "timestamp": "2024-01-29T00:00:00.000Z",
+    "tokens": [
+      {
+        "networkId": 200,
```
Comment (ebma) [inline packages/token-lists/lists/pendulum.json]: Should be `300` for all of them.
```suggestion
        "networkId": 300,
```

File: packages/ui/icons/network/circle/PendulumCircle.tsx [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import * as React from 'react'
+
+// TODO change
```
Comment (ebma) [inline packages/ui/icons/network/circle/PendulumCircle.tsx]: Change to what?

File: packages/ui/icons/network/naked/PendulumNaked.tsx [hunk 9]
```diff
@@ -0,0 +1,13 @@
+import * as React from 'react'
+
+//TODO change
```
Comment (ebma) [inline packages/ui/icons/network/naked/PendulumNaked.tsx]: Are we still missing the icons?

File: packages/ui/icons/SubwalletIcon.tsx
Comment (ebma) [inline packages/ui/icons/SubwalletIcon.tsx:1]: Let's undo these changes maybe.

File: apps/pool/lib/hooks/useTokensFromPair.ts [hunk 10]
```diff
@@ -31,7 +31,7 @@ export function useTokensFromPair(pair: Pair) {
       new Token({
         address: pair.id.includes(':') ? pair.id.split(':')[1] : pair.id,
         name: 'Zenlink LP Token',
-        decimals: pair.chainId === ParachainId.AMPLITUDE ? 12 : 18,
+        decimals: (pair.chainId === ParachainId.AMPLITUDE || pair.chainId === ParachainId.PENDULUM ) ? 12 : 18,
```
Comment (ebma) [inline apps/pool/lib/hooks/useTokensFromPair.ts:34]: Would it make sense to extract this `(pair.chainId === ParachainId.AMPLITUDE || pair.chainId === ParachainId.PENDULUM )` check to an extra function and replace all occurences? Having this condition everywhere might be a little too much.

File: packages/config/router/src/index.ts (hunk 3 above)
Comment (ebma) [reply packages/config/router/src/index.ts]: This means we might need to change some other occurrences where we assume that Pendulum has `12` decimals for every asset.

File: packages/ui/icons/network/naked/PendulumNaked.tsx (hunk 9 above)
Comment (gianfra-t) [reply packages/ui/icons/network/naked/PendulumNaked.tsx]: Yes and no. The svg from [here ](https://pendulumchain.org/brand-assets) had a different format to how the ones from amplitude was defined. It works, but I was unsure they are correct from the design standpoint.

File: packages/ui/icons/SubwalletIcon.tsx
Comment (gianfra-t) [reply packages/ui/icons/SubwalletIcon.tsx:1]: But these changes were introduced in [this](https://github.com/pendulum-chain/zenlink-interface/pull/2/commits/3a76ad045e2f4763480a41e2c9b07de5cdf14537) commit, not by us. They must come from when i merged with the upstream branch. I should have rebased sorry.

File: packages/chain/parachains.json (hunk 1 above)
Comment (gianfra-t) [reply packages/chain/parachains.json]: yes, you are right, bifrost is using 300.

File: packages/config/router/src/index.ts (hunk 3 above)
Comment (gianfra-t) [reply packages/config/router/src/index.ts]: You are right, I assumed that every asset had 12 decimals. Let me double check.

File: packages/graph-client/src/appolo/client.ts (hunk 4 above)
Comment (gianfra-t) [reply packages/graph-client/src/appolo/client.ts:67]: You mean in archives? Yes, probably.

File: apps/pool/lib/hooks/useTokensFromPair.ts (hunk 10 above)
Comment (gianfra-t) [reply apps/pool/lib/hooks/useTokensFromPair.ts:34]: I didn't think about it. It could be, I guess it is a matter of decision. The only "problem" I see is that we would improve readability in the condition, but add too many imports.
On the other hand, this condition will always be the same, no other chains will ever share the handlers. 

I have no bias towards any option, what do you prefer?

File: packages/currency/src/constants/tokenAddresses.ts
Comment (gianfra-t) [reply packages/currency/src/constants/tokenAddresses.ts:1]: Added `USDT` and `USDC` also. With the same logic should we also add `XLM` and the other assets that are not here? Not sure why `GLMR` is not here already.

File: packages/ui/icons/SubwalletIcon.tsx
Comment (ebma) [reply packages/ui/icons/SubwalletIcon.tsx:1]: I see, thanks for pointing it out 👍

File: apps/pool/lib/hooks/useTokensFromPair.ts (hunk 10 above)
Comment (ebma) [reply apps/pool/lib/hooks/useTokensFromPair.ts:34]: That's a fair point. Let's leave it as is then, the condition is long but it's readable. If the Zenlink team wants something different, they'll let us know anyway.

File: packages/currency/src/constants/tokenAddresses.ts
Comment (ebma) [reply packages/currency/src/constants/tokenAddresses.ts:1]: Where exactly would you want to add XLM and GLMR? To a `Record`?

File: packages/ui/icons/network/naked/PendulumNaked.tsx (hunk 9 above)
Comment (ebma) [reply packages/ui/icons/network/naked/PendulumNaked.tsx]: Which logos did you use (ie. what's the name of the files in the brand assets folder)?

File: packages/currency/src/constants/tokenAddresses.ts
Comment (gianfra-t) [reply packages/currency/src/constants/tokenAddresses.ts:1]: Yes I was thinking to a Record, but it would have to be a new one. What is strange to me is that `GLMR` for instance is used in the UI already, but has not been added here. So it must not be mandatory.

File: packages/ui/icons/network/naked/PendulumNaked.tsx (hunk 9 above)
Comment (gianfra-t) [reply packages/ui/icons/network/naked/PendulumNaked.tsx]: I used the ones in Brand Assets, `LOGO/SVG/Pendulum - RGB_Logomark - dark on bright` for `PendulumCircle` and `Pendulum - RGB_Logomark - dark` for the `Naked` one.

And then after modifications of the SVG so it would be compatible with the format used in this repository, I checked using a [visualizer](https://www.svgviewer.dev/) that the logos looked the same. It seems to me they do look the same.

Comment (ebma) [reply packages/ui/icons/network/naked/PendulumNaked.tsx]: I see. As far as I can tell, the 'circle' icons all have an some extra circle-shaped background. I think the 'LOGO/SVG/Pendulum - RGB_Logomark - dark on bright' is indeed the correct fit but we somehow need to make the background be circle shaped instead of rectangular. I checked the files you added using the visualizer you linked to, and it seems to me like the 'naked' and 'circle' logo we have now are the same (?).
We can reach out to Klaus to quickly provide us with this custom icon.

Besides that, we still need to pass the `{...props}` into the svg like all the other icons do.

Comment (gianfra-t) [reply packages/ui/icons/network/naked/PendulumNaked.tsx]: > it seems to me like the 'naked' and 'circle' logo we have now are the same

I must have removed the background while attempting to remove some errors. I think is better reach out to Klaus.

File: packages/currency/src/constants/tokenAddresses.ts
Comment (ebma) [reply packages/currency/src/constants/tokenAddresses.ts:1]: I am not totally sure what they use all these record mappings for but I guess we don't really need it for GLMR and XLM.

File: packages/ui/icons/network/circle/PendulumCircle.tsx (hunk 8 above)
Comment (gianfra-t) [reply packages/ui/icons/network/circle/PendulumCircle.tsx]: This one has been modified now!

**Review by ebma — APPROVED** [review]
I didn't test it locally but it looks good to me 👍  let's close and not merge this PR then and open up a new PR from this fork to the upstream repository. We should then mention it in the related issue ticket. 

Great job by the way @gianfra-t 👌

**Comment by gianfra-t** [comment]
The test I did locally worked (with the old svg's) but there is a caveat, I had to use amplitude's rpc when building (or the local node), otherwise you get an error regarding the missing extrinsics! But after everything is configured in Pendulum, this should work.

**Comment by ebma** [comment]
That's not too bad. Last time, for the other PR to integrate Amplitude into the Zenlink UI, the Zenlink team also disabled the network from being selectable for the time being until the pallet configuration etc hit production. And once ready, they enabled that flag again. Was quite easy, I guess they can do something similar again.
